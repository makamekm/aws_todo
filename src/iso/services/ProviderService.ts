import { GraphQLClient } from "graphql-hooks";

export class ProviderService {
    public client: GraphQLClient;
}
