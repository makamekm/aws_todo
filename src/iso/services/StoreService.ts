export class StoreService {
    [key: string]: any;
}
