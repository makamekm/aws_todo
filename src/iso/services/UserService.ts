import { inject } from "react-ioc";
import { StoreService } from "./StoreService";

export class UserService {
    @inject public store: StoreService;

    public get user() {
        return this.store.store$.value.user;
    }
}
