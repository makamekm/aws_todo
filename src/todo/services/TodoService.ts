import { useMutation, useQuery } from "graphql-hooks";
import * as moment from "moment";
import * as React from "react";
import { inject } from "react-ioc";
import { BehaviorSubject, from, Subject } from "rxjs";
import { useObservable } from "rxjs-hooks";
import { map, switchMap, tap, toArray } from "rxjs/operators";
import { StoreService } from "../../iso/services/StoreService";

export const QUERY_TODO_LIST = `query ListTodo($year: Int!, $month: Int!, $day: Int!) {
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;
export const QUERY_TODO_STATS = `query StatsTodo($year: Int!, $month: Int!) {
    todoQuery {
        stats(
            year: $year,
            month: $month,
        ) {
            year
            month
            day
            finished
            total
            unfinished {
                id
                name
                isDone
            }
        }
    }
}`;

export const MUTATION_TODO_CREATE = `mutation CreateTodo(
        $year: Int!,
        $month: Int!,
        $day: Int!,
        $name: String!
    ) {
    todoMutation {
        create(
            year: $year,
            month: $month,
            day: $day,
            name: $name,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;

export const MUTATION_TODO_DELETE = `mutation DeleteTodo(
        $id: ID!,
        $year: Int!,
        $month: Int!,
        $day: Int!,
    ) {
    todoMutation {
        delete(id: $id) {
            id
            name
            isDone
            date
            updated
        }
    }
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;

export const MUTATION_TODO_TOGGLE = `mutation ToggleTodo(
    $id: ID!,
    $year: Int!,
    $month: Int!,
    $day: Int!,
) {
    todoMutation {
        toggle(id: $id) {
            id
            name
            isDone
            date
            updated
        }
    }
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;

export const MUTATION_TODO_EDIT = `mutation EditTodo(
        $id: ID!,
        $name: String!,
        $year: Int!,
        $month: Int!,
        $day: Int!,
    ) {
    todoMutation {
        edit(id: $id, name: $name) {
            id
            name
            isDone
            date
            updated
        }
    }
    todoQuery {
        list(
            year: $year,
            month: $month,
            day: $day,
        ) {
            id
            name
            isDone
            date
            updated
        }
    }
}`;

export interface ITodo {
    id: number;
    name: string;
    isDone: boolean;
    date: string;
    updated: string;
}

export class TodoService {
    @inject public storeService: StoreService;
    public searchString$ = new BehaviorSubject("");
    public data$: BehaviorSubject<ITodo[]> = new BehaviorSubject(null);
    public mutationError$: Subject<string> = new Subject();
    public queryErrors$: BehaviorSubject<string[]> = new BehaviorSubject([]);
    public date$ = new BehaviorSubject(moment(this.storeService.date));

    public findStats(stats, year, month, day) {
        return stats && stats.find((s) => s.year === year && s.month === month && s.day === day);
    }

    public useStatsService = () => {
        const response = useQuery(QUERY_TODO_STATS, {
            variables: {
                year: this.date$.value.year(),
                month: this.date$.value.month(),
            },
        });
        const { data: queryData, loading, refetch, error } = response;
        const stats = queryData && queryData.todoQuery && queryData.todoQuery.stats;
        useObservable(() => this.date$.pipe(
            switchMap(refetch),
        ));

        if (error) {
            this.pipeQueryErrors(response);
        }

        return {
            stats,
            loading,
            refetch,
            getStats: (date) => this.findStats(
                stats,
                date.year(),
                date.month(),
                date.date(),
            ),
        };
    }

    public useDataService = () => {
        const response = useQuery(QUERY_TODO_LIST, {
            variables: {
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        const { data: queryData, loading, refetch, error } = response;
        const list = queryData && queryData.todoQuery && queryData.todoQuery.list;

        if (error) {
            this.pipeQueryErrors(response);
        }

        if (!this.data$.value && list) {
            this.data$.next(list);
        }

        React.useEffect(() => {
            if (list) {
                this.data$.next(list);
            }
        }, [list]);

        return {
            refetch,
            loading,
        };
    }

    public useItemService = (item) => {
        const [deleteTodo, { loading: deleteTodoLoading }] = useMutation(MUTATION_TODO_DELETE);
        const [toggleTodo, { loading: toggleTodoLoading }] = useMutation(MUTATION_TODO_TOGGLE);
        const [editTodo, { loading: editTodoLoading }] = useMutation(MUTATION_TODO_EDIT);

        return {
            loading: deleteTodoLoading || toggleTodoLoading || editTodoLoading,
            deleteTodo: () => this.deleteTodo(deleteTodo, item.id),
            toggleTodo: () => this.toggleTodo(toggleTodo, item.id),
            editTodo: (name) => this.editTodo(editTodo, item.id, name),
        };
    }

    public useCreateService = () => {
        const [createTodo, { loading: createTodoLoading }] = useMutation(MUTATION_TODO_CREATE);

        return {
            loading: createTodoLoading,
            createTodo: () => this.createTodo(createTodo, this.searchString$.value),
        };
    }

    public removeQueryError(index: number) {
        const list = Array.from(this.queryErrors$.value);
        list.splice(index, 1);
        this.queryErrors$.next(list);
    }

    private pipeMutationErrors(response) {
        from(response.graphQLErrors).pipe(
            map((error: any) => error.message),
            tap((error: string) => this.mutationError$.next(error)),
        ).subscribe();
    }

    private pipeQueryErrors(response) {
        from(response.graphQLErrors).pipe(
            map((error: any) => error.message),
            toArray(),
            tap((errors: string[]) => this.queryErrors$.next(errors)),
        ).subscribe();
    }

    private async createTodo(createTodo, name) {
        const result = await createTodo({
            variables: {
                name,
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        if (!result.error) {
            this.data$.next(result.data.todoQuery.list);
            this.searchString$.next("");
        } else {
            this.pipeMutationErrors(result);
        }
    }

    private async deleteTodo(deleteTodo, id) {
        const result = await deleteTodo({
            variables: {
                id,
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        if (!result.error) {
            this.data$.next(result.data.todoQuery.list);
            this.searchString$.next("");
        } else {
            this.pipeMutationErrors(result);
        }
    }

    private async editTodo(editTodo, id, name) {
        const result = await editTodo({
            variables: {
                id,
                name,
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        if (!result.error) {
            this.data$.next(result.data.todoQuery.list);
            this.searchString$.next("");
        } else {
            this.pipeMutationErrors(result);
        }
    }

    private async toggleTodo(toggleTodo, id) {
        const result = await toggleTodo({
            variables: {
                id,
                year: this.date$.value.year(),
                month: this.date$.value.month(),
                day: this.date$.value.date(),
            },
        });
        if (!result.error) {
            this.data$.next(result.data.todoQuery.list);
            this.searchString$.next("");
        } else {
            this.pipeMutationErrors(result);
        }
    }
}
