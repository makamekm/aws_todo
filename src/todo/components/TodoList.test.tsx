import * as React from "react";
import { inject } from "react-ioc";
import {cleanup} from "react-testing-library";
import {renderAndWrapWithClient} from "../../test/Client";
import { TodoService } from "../services/TodoService";
import { getClientTodo } from "../test/Client";
import {
    TodoList,
} from "./TodoList";

describe("Todo", () => {
    describe("Components", () => {
        describe("TodoList", () => {
            afterEach(cleanup);

            it("should match snapshot", async () => {
                const client = getClientTodo();
                const html = await renderAndWrapWithClient(
                    <TodoList/>,
                    client,
                );
                expect(html.container).toMatchSnapshot();
            });

            it("should ServiceTodo exist", async () => {
                const client = getClientTodo();
                const isTodoServiceExist = jest.fn();
                class TodoServiceCheck extends React.PureComponent<{}> {
                    @inject private todoService: TodoService;

                    public render() {
                        if (this.todoService) {
                            isTodoServiceExist();
                        }

                        return (
                            <div/>
                        );
                    }
                }
                await renderAndWrapWithClient(
                    <TodoList>
                        <TodoServiceCheck/>
                    </TodoList>,
                    client,
                );
                expect(isTodoServiceExist).toBeCalled();
            });
        });
    });
});
