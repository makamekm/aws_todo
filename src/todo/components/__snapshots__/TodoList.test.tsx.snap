// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Components TodoList should match snapshot 1`] = `
<div>
  <div
    class="ant-list ant-list-sm ant-list-split ant-list-bordered"
  >
    <div
      class="ant-list-header"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <input
            class="ant-input"
            placeholder="Here is your today's todo list"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="ant-spin-nested-loading"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-list-item"
            >
              <div
                class="ant-list-item-meta"
              >
                <div
                  class="ant-list-item-meta-content"
                >
                  <h4
                    class="ant-list-item-meta-title"
                  >
                    <div>
                      <span
                        class="ant-typography"
                      >
                        todo1
                        <button
                          aria-label="edit"
                          class="ant-typography-edit"
                          style="border: 0px; background: transparent; padding: 0px; line-height: inherit;"
                        >
                          <i
                            aria-label="icon: edit"
                            class="anticon anticon-edit"
                            role="button"
                          >
                            <svg
                              aria-hidden="true"
                              class=""
                              data-icon="edit"
                              fill="currentColor"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 0 0 0-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 0 0 9.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                              />
                            </svg>
                          </i>
                        </button>
                      </span>
                    </div>
                  </h4>
                </div>
              </div>
              <ul
                class="ant-list-item-action"
              >
                <li>
                  <a>
                    toggle
                  </a>
                  <em
                    class="ant-list-item-action-split"
                  />
                </li>
                <li>
                  <a
                    href="javascript:;"
                    style="color: red;"
                  >
                    remove
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-list-item"
            >
              <div
                class="ant-list-item-meta"
              >
                <div
                  class="ant-list-item-meta-content"
                >
                  <h4
                    class="ant-list-item-meta-title"
                  >
                    <div>
                      <span
                        class="ant-typography"
                      >
                        todo2
                        <button
                          aria-label="edit"
                          class="ant-typography-edit"
                          style="border: 0px; background: transparent; padding: 0px; line-height: inherit;"
                        >
                          <i
                            aria-label="icon: edit"
                            class="anticon anticon-edit"
                            role="button"
                          >
                            <svg
                              aria-hidden="true"
                              class=""
                              data-icon="edit"
                              fill="currentColor"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 0 0 0-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 0 0 9.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                              />
                            </svg>
                          </i>
                        </button>
                      </span>
                    </div>
                  </h4>
                </div>
              </div>
              <ul
                class="ant-list-item-action"
              >
                <li>
                  <a>
                    toggle
                  </a>
                  <em
                    class="ant-list-item-action-split"
                  />
                </li>
                <li>
                  <a
                    href="javascript:;"
                    style="color: red;"
                  >
                    remove
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
